### ISSUE TITLE
Three proposals to improve your software

### ISSUE THREAD

**Issue by Testostheron1963-de** [DexcTrack/dexctrack#8]
Hello,

1.)

I think, I have an idea to set the "desired range" values of the diagram depending on settings of the Dexcom G6 device. I think it would be better to set the diagrams "desired range" to the warning levels, if the warning levels are enabled. This would be better for user, who have special levels for these settings prescribed by the doctor to evaluate the diagram.

2.)

I have to use two Insulin types, one with a long period of work (for example "Levemir") and one withe a short period of work (for example "Humalog"). In your diagram all Insulin injections have the same view. That means, you can't distinguish between these two kinds of Insulin injections. Wouldn't it be better to draw these two kinds of Insulin (long working period and short working period) using different colours ?

3.)

In your diagram the value of the injected Insulin is decribed using 2 decimals. I think, nobody can adjust the amount of Insulin that has to be injected in 1/100 of an Insulin unit. In my opinion this makes no sense at all.

**Comment by DexcTrack** [comment]
I've implemented a simple change to address proposal 3. That was a good idea, and easy to fix. The Dexcom receiver device allows the user to specify fractions of insulin units in multiples of .05, so I wanted to preserve the option to print those values out. By switching the print statement to use '%g', trailing 0 digits after the decimal point will no longer be displayed. If the user specifies a whole unit, '.00' will not appear.

**Comment by DexcTrack** [comment]
For proposal 2, I'd love to be able to distinguish the insulin type, but the Dexcom Receiver doesn't provide any way to do so. When you enter an Event of type Insulin, you can only specify the number of units. To implement your proposal, Dexcom would have to update their firmware to add an option to indicate the insulin type. 

I use the Insulin Event to specify **adjustments** to my standard dose. If my standard dosage for morning insulin is 10 units of short period insulin and 15 units of long period insulin, the only time I add an Insulin Event is when I have to add to the short period insulin dosage to reduce a temporary high blood glucose level. I specify the number of additional short term insulin units.

I would like to add some way of specifying the standard dosage levels and display that info in the graph, but I haven't figured out a good way to do that yet. I'm open to suggestions.

**Comment by Testostheron1963-de** [comment]
Thank you for your fast answer !

For point 3:

If I take in your point of view, you are right.

For point 2:

I think, you're using an Dexcom G4/G5 device. If you do that, you can only enter an amount of Insulin units without the ability to specify the Insulin type. 

Is my assumption true in your case ?

**BUT: This is not the case when using an Dexcom G6 device. When doing so, you can input the amount of Insulin units for each of these Insulin types. I know, that this is a new feature provided by the Dexcom G6 device. And this is my case of usage.**

If you use the original software, called Dexcom Studio, you can see that the Dexcom application is able to determine the Insulin type. If you use the recommended app for Apple devices, it's the same procedure. So it should be possible to distinguish between these 2 Insulin types.
My doctor uses the professional Dexcom app to read all the values out of the Dexcom device without any connection to the Dexcom cloud and you can see that the Insulin dosage is clear separated between these two types of Insulin. I have to inject 10 units of long period Insulin
and 8 units of of short period Insulin in the morning and the same amount of Insulin in the evening. But when the it goes from winter to spring I have to change the amount of long period Insulin. So it wood be good to display that visually on the screen.

I would be able to send you my database file for research purposes.

**Are you able to implement my first proposal ?**

Thank you for your time and your work !

**Comment by Testostheron1963-de** [comment]
I would like to do such program changes by myself but I'm a software developer for embedded devices using hardware closed programming in C / C++ / Assembler. I just started out to create Python programs and so I'm far away to be an expert in this range. But for me it's a good point
of entry to have a look at the programming techniques for Python.

Thank you for your help !

**Comment by DexcTrack** [comment]
My impression is that the G5 and G6 Receiver hardware is identical. But I don't know for sure. I originally bought a G5 Receiver, which I later "upgraded" to G6. My current Receiver has Part Number **MT22949**, Part Revision **013**, Firmware version **5.1.1.022**, SW version **SW11269**.

The Dexcom website does not make it easy to figure out what the latest Firmware or Software release is. All I can find is instructions on how to convert a G5 system to a G6 system. I can't find anything to tell me if there is a Firmware version newer than **5.1.1.022**, or a Software version newer than **SW11269**. What information do you see on your G6 Receiver when you run **Menu -> Settings -> Receiver Info**?

It looks like the Event sub-type is used to specify 'Fast Acting' or 'Long Acting' insulin types for devices which allow the user to pick one. I'll merge in an update to print out "Fast Insulin" or "Long Insulin" based off of this sub-type. I'll have to guess at the enumeration values. On my Receiver, the sub-type is always 0 which translates to 'None'. I'm going to guess 1 = 'Fast' and 2 = 'Long'. If these are backwards, let me know.

**Comment by Testostheron1963-de** [comment]
Your impression is definitely wrong. For example the Dexcom G6 has only a Touchscreen on it to select the functions to handle it except of one Pushbutton below the display to activate the device from sleep - mode. Also the package of such an device is totally different from that of an Dexcom G5. So there is no way to upgrade a Dexcom G5 device to a Dexcom G6 device. The Dexcom G6 is designed to work in Portrait mode, The Dexcom G5 is designed to work in Landscape mode whereby the buttons to handle this device are located at the right side. 

I have added the images of these two devices so you can see the difference between them.

Can you see the difference

![Dexcom-G6](https://user-images.githubusercontent.com/55926107/74821117-69027f00-5303-11ea-8099-a0fb8be6c8d6.jpg)
![Dexcom-G5](https://user-images.githubusercontent.com/55926107/74821132-6ef86000-5303-11ea-8eb8-131f94cedcb8.jpg)

Additionally these two devices have different sensor systems and different transmitting devices.

Firmware version: 5.1.3.026
Software version: SW11713

**Comment by Testostheron1963-de** [comment]
Thank you so much for your work on this improvement of your software.
It works great !

I have another idea for your software. It depends on my initial comment (point 1).
I will describe this feature in more detailed at a later time.

Additionally I have an idea of another improvement of your software,
It is depending on the way, I can document all the values, that have been stored in your database.

But currently I have to do other things.
But I will return soon to help you to improve your software.

**Comment by DexcTrack** [comment]
My "G5" receiver device (a touchscreen with one button) is identical to your "G6" receiver device. I upgraded the software on it to provide it support for G6 transmitters and sensors. However, my software version is older than what you have. Dexcom does a poor job of documenting their software versions. I didn't even know that any newer SW version was available. Maybe, I'll try updating my version. Before I do so, I'd like to know what modifications have been made, and whether I can revert to the current version if I don't like the some of those modifications.

I'm not sure about your Proposal 1. I want my target range to be 75 - 200. On my receiver I've set my High Alert level to 300, because I don't want to suffer annoying alarm sounds when I barely eclipse 200. If I followed your proposal, then my Target range would be 75 - 300, which is way too wide. I built in a capability for the user to set their desired Target Range, by clicking the "Set New Target Range" button in the bottom right corner of the screen. So, if a user has set their own desirable Target Range, how would that be handled with your proposal? Maybe I could default the initial Target Range to (Low Alert, High Alert) the first time the user runs my program. Would that be sufficient? If the user modifies the High Alert level later on, should the Target Range change, or should we keep the value which they set using the "Set New Target Range" button?

Thank you for your proposals. I appreciate your input.
